File: elementary_data_structures/python/stack.py
class StackEmptyException(Exception):
    pass


class Element:
    __slots__ = ["data", "next"]

    def __init__(self, data):
        self.data = data
        self.next = None


class Stack:
    __slots__ = ["size", "head"]

    def __init__(self):
        self.size = 0
        self.head = None

    def push(self, data):
        self.size += 1

        z = Element(data)
        z.next = self.head
        self.head = z

    def pop(self):
        if self.size == 0:
            raise StackEmptyException

        self.size -= 1

        popped = self.head
        result = popped.data
        self.head = popped.next

        return result

    def print_stack(self):
        print(f"size = {self.size} : ", end="")

        if not self.head:
            print("(empty)")

            return

        p = self.head
        while p:
            print(p.data, end=" -> " if p.next else "\n")

            p = p.next


if __name__ == "__main__":
    s = Stack()

    while True:
        print("1:push 2:pop 3:print > ", end="")
        if (op := int(input())) == 1:
            print("input data > ", end="")
            data = int(input())

            s.push(data)
        elif op == 2:
            try:
                popped = s.pop()
                print(f"popped = {popped}")
            except StackEmptyException:
                print("stack is empty")
        elif op == 3:
            s.print_stack()
        else:
            print("invalid operation")

File: elementary-data-structures/golang/stack.go
package main

import "fmt"

type Element struct {
	Data int
	Next *Element
}

type Stack struct {
	Size int
	Head *Element
}

func (S *Stack) push(e *Element) {
	S.Size++

	e.Next = S.Head
	S.Head = e
}

func (S *Stack) pop() (int, bool) {
	if S.Head == nil {
		return 0, false
	}

	S.Size--

	popped := S.Head
	result := popped.Data
	S.Head = popped.Next

	return result, true
}

func (S *Stack) printStack() {
	fmt.Printf("size = %d : ", S.Size)

	if S.Head == nil {
		fmt.Println("(empty)")

		return
	}

	for p := S.Head; p != nil; p = p.Next {
		fmt.Print(p.Data)
		if p.Next != nil {
			fmt.Print(" -> ")
		} else {
			fmt.Print("\n")
		}
	}
}

func main() {
	S := &Stack{}

	for {
		fmt.Print("1:push 2:pop 3:print > ")
		var op int
		fmt.Scanf("%d", &op)

		switch op {
		case 1:
			fmt.Print("input data > ")
			var data int
			fmt.Scanf("%d", &data)
			S.push(&Element{data, nil})
		case 2:
			popped, empty := S.pop()
			if !empty {
				fmt.Println("stack is empty")
			} else {
				fmt.Printf("popped data = %d\n", popped)
			}
		case 3:
			S.printStack()
		default:
			fmt.Println("invalid operation")
		}
	}
}
